encrypt and decrypt functions

Enmk opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New functions `encrypt`, `aes_encrypt_mysql`, `decrypt`, `aes_decrypt_mysql`. These functions are working slowly (below ClickHouse standards), so we consider it as an experimental feature.


Detailed description / Documentation draft:
New encryption/decryption functions

## encrypt(mode, plaintext, key, [iv, aad])
## aes_encrypt_mysql(mode, plaintext, key, [iv])
## decrypt(mode, ciphertext, key, [iv, aad])
## aes_decrypt_mysql(mode, ciphertext, key, [iv]). 
Support encrypting/decrypting in following modes:

* aes-128-ecb
* aes-192-ecb
* aes-256-ecb
* aes-128-cbc
* aes-192-cbc
* aes-256-cbc
* aes-128-cfb1
* aes-192-cfb1
* aes-256-cfb1
* aes-128-cfb8
* aes-192-cfb8
* aes-256-cfb8
* aes-128-cfb128
* aes-192-cfb128
* aes-256-cfb128
* aes-128-ofb
* aes-192-ofb
* aes-256-ofb

Modes supported only by `encrypt`/`decrypt`:
* aes-128-gcm
* aes-192-gcm
* aes-256-gcm

`aes_encrypt_mysql`/`aes_decrypt_mysql` functions are to provide compatibility with MySQL when encrypting/decrypting data. That is: do not check `iv` length, and also "fold" keys that are longer than required by the cipher.
For '-gcm' modes, an `AEAD` tag is appended to the resulting ciphertext according to the RFC5116 on encryption and is expected to be present there on decryption.
AAD-parameter is taken into account only for '-gcm' modes.
IV is required for '-gcm' modes, and can be optional for other modes.
...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Enmk committed:
encrypt, aes_encrypt_mysql, decrypt, aes_decrypt_mysql functions

Functions to encrypt/decrypt any input data with OpenSSL's ciphers
with custom key, iv, and add (-gcm mode only).

_mysql versions are 100% compatitable with corresponding MySQL functions

Supported modes depend on OpenSSL version, but generally are:
aes-{128,192,56}-{ecb,cbc,cfb1,cfb8,cfb128,ofb,gcm}

Please note that in a -gcm mode a 16-byte tag is appended to the ciphertext
on encryption and is expected to be found at the end of ciphertext on decryption.

Added tests that verify compatibility with MySQL functions,
and test vectors for GCM mode from OpenSSL.

Added masking rules for aes_X funtions
Rules are installed by default to config.d/query_masking_rules.xml

Commit:
Adding AES encryption tests in TestFlows.

abyss7 commented on programs/server/CMakeLists.txt:
```diff
@@ -31,3 +31,4 @@ if (GLIBC_COMPATIBILITY)
 endif ()
 
 install(FILES config.xml users.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server COMPONENT clickhouse)
+install(FILES config.d/query_masking_rules.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server/config.d COMPONENT clickhouse)
```
Looks like this file (or file with the same name) is already installed and symlinked to config.d folder.

That's the reason why fast-test is failing.

Enmk replied:
My impression is that testing tools should provide workarounds for issues like this, not production builds. BTW, it should be fixed now in #13382

Enmk committed:
Fixed AES encrypt/decrypt performance

Improved execution time of EVP_DecryptInit_ex/EVP_EncryptInit_ex with some hacks
Fixed all-const case

Enmk committed:
Fixed compilation for older (pre-3.0.0) OpenSSL versions

Enmk committed:
Excluding tests from fasttest runs since they depend on external libraries: OpenSSL

Enmk committed:
Fixed data race caused by EVP_CIPHER_fetch()

Commit:
Updating TestFlows to 1.6.45 to fix stability issues with multi line commands.

Enmk committed:
Fixed Yandex-specific builds :(

qoega commented on debian/clickhouse-server.install:
```diff
@@ -2,5 +2,6 @@ usr/bin/clickhouse-server
 usr/bin/clickhouse-copier
 usr/bin/clickhouse-report
 etc/clickhouse-server/config.xml
+etc/clickhouse-server/config.d/*.xml
```
Why you don't add new configuration to default config.xml? We try not to bring anything to production via  `config.d` and users can use `config.d` safely without conflicts with our packages

qoega commented on programs/server/CMakeLists.txt:
```diff
@@ -29,6 +29,8 @@ set (CLICKHOUSE_SERVER_LINK
 clickhouse_program_add(server)
 
 install(FILES config.xml users.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server COMPONENT clickhouse)
+install(FILES config.xml users.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server COMPONENT clickhouse)
```
It is duplicate?

Enmk replied:
yep

qoega commented on programs/server/CMakeLists.txt:
```diff
@@ -29,6 +29,8 @@ set (CLICKHOUSE_SERVER_LINK
 
 install(FILES config.xml users.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server COMPONENT clickhouse)
+install(FILES config.xml users.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server COMPONENT clickhouse)
+install(FILES config.d/query_masking_rules.xml DESTINATION ${CLICKHOUSE_ETC_DIR}/clickhouse-server/config.d COMPONENT clickhouse)
```
Same as above. Probably it is better to add query_masking_rules directly to config.xml and don't touch config.d.

qoega commented on src/Functions/FunctionHelpers.cpp:
```diff
@@ -137,7 +137,7 @@ void validateArgumentsImpl(const IFunction & func,
         const auto descriptor = descriptors[i];
         if (int error_code = descriptor.isValid(arg.type, arg.column); error_code != 0)
-            throw Exception("Illegal type of argument #" + std::to_string(i)
+            throw Exception("Illegal type of argument #" + std::to_string(argument_offset + i + 1) // +1 is for human-friendly 1-based indexing
```
Better to add this comment directly to exception message.

qoega commented on tests/testflows/aes_encryption/configs/clickhouse/config.xml:
```diff
@@ -0,0 +1,436 @@
+<?xml version="1.0"?>
```
Please do not copy/pase default config.xml. It makes impossible to enable features by default in all tests and prod. It is better to organize config as default one plus small differences. Such behavior is used in integration tests.

vzakaznikov replied:
I agree that this can be structured in a better way. However, this would affect everything in the `tests/testflows` folder. This  will be done in separate PR.

qoega commented on tests/testflows/aes_encryption/configs/clickhouse/users.xml:
```diff
@@ -0,0 +1,133 @@
+<?xml version="1.0"?>
```
Same as config.xml.

vzakaznikov replied:
This will be fixed in another PR.

qoega commented on tests/testflows/aes_encryption/docker-compose/clickhouse-service.yml:
```diff
@@ -0,0 +1,28 @@
+
+services:
+  clickhouse:
+    image: yandex/clickhouse-integration-test
```
It is better to move all docker compose files to testflows image. It makes possible to update images inside same PR and not use latest version.
It is better to pass tag directly from runner to `yandex/clickhouse-integration-test` via env. It is possible to add dependency testflows<-integration to CI and your task will receive tag from PR build.

vzakaznikov replied:
The same here. Let's move this to be part of restructuring `test/testflows` folder in a separate PR.

qoega commented on tests/testflows/aes_encryption/tests/common.py:
```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+modes = [
```
is it generated manually? is it possible to check that all modes are listed here?

vzakaznikov replied:
These modes are defined in the Software Requirements Specification (SRS). I will add the source in Markdown.

Enmk committed:
Moved default values for query_masking rules for encrypt/decrypt to config.xml

Enmk commented:
@qoega Fixed all notes regarding install paths and default configurations.

Enmk committed:
Moving to TestFlows version 1.6.46

vzakaznikov committed:
Adding source for the SRS.

alexey-milovidov committed:
Delete gtest_openssl.cpp

alexey-milovidov commented on base/common/defines.h:
```diff
@@ -70,10 +70,12 @@
 #    define NO_SANITIZE_UNDEFINED __attribute__((__no_sanitize__("undefined")))
 #    define NO_SANITIZE_ADDRESS __attribute__((__no_sanitize__("address")))
 #    define NO_SANITIZE_THREAD __attribute__((__no_sanitize__("thread")))
+#    define NO_SANITIZE_MEMORY __attribute__((__no_sanitize__("memory")))
```
It is very difficult to use. It will work only if function reads uninitialized memory and writes nothing. I will remove to avoid confusion.

alexey-milovidov committed:
Update defines.h

alexey-milovidov commented on programs/server/config.d/query_masking_rules.xml:
```diff
@@ -15,5 +15,14 @@
             <replace>[hidden]</replace>
         </rule>
+
+        <rule>
```
No need to duplicate.

alexey-milovidov committed:
Update query_masking_rules.xml

alexey-milovidov committed:
Update ColumnString.cpp

alexey-milovidov committed:
Update FunctionsAES.cpp

alexey-milovidov reviewed (commented):
It's better to add some performance tests.

alexey-milovidov committed:
Update FunctionsAES.h

alexey-milovidov committed:
Update FunctionsAES.h

alexey-milovidov committed:
Merge branch 'master' into AES_encrypt_decrypt

alexey-milovidov commented:
```
clickhouse-benchmark <<< "SELECT DISTINCT Title = decrypt('aes-128-cbc', encrypt('aes-128-cbc', Title, 'key-key-key-key-'), 'key-key-key-key-') FROM test.hits"
```

16 cores and 100 MB/sec.

```
OSSL_PARAM_locate
EVP_CipherInit_ex
__pthread_rwlock_rdlock !!!
OPENSSL_cleanse
```

Not quite acceptable.

alexey-milovidov commented:
```
SELECT DISTINCT Title = decrypt('aes-128-cbc', encrypt('aes-128-cbc', Title, 'key-key-key-key-'), 'key-key-key-key+')
FROM test.hits

→ Progress: 131.07 thousand rows, 10.46 MB (1.27 million rows/s., 101.30 MB/s.)  1%
Received exception from server (version 20.10.1):
Code: 454. DB::Exception: Received from localhost:9000. DB::Exception: Failed to decrypt. OpenSSL error code: 956301412.
```

Poor usability of the error message (the error code looks cryptic).

Enmk committed:
Empty string optimization for encryption

alexey-milovidov committed:
Merge branch 'master' into AES_encrypt_decrypt

Enmk committed:
Performance tests for encrypt and decrypt functions

alexey-milovidov committed:
Merge with master

Enmk committed:
Fixed argument checks and updated tests accordingly

Enmk committed:
Merge remote-tracking branch 'upstream/master' into AES_encrypt_decrypt

Enmk committed:
AES: Attempt to make performance tests faster and more stable.

alexey-milovidov commented:
It's strange to have GCM but not to have CTR mode.

alexey-milovidov commented:
After porting to BoringSSL it has "magically" become 30 times faster: from 100 MB/sec to 3000 MB/sec.

alexey-milovidov commented:
`aes-128-ctr` is also supported automatically.

alexey-milovidov commented:
I consider this PR as finally done. It works perfectly.
